### DISCUSSION TITLE
Is it possible to do SPICE simulation and generate an SVG from the same netlist?

### DISCUSSION THREAD

**Discussion by gsingh93** [xesscorp/skidl#134] [Q&A] [answered]
I've been able to run a SPICE simulation using `skidl.pyspice`, and separately I've also been able to generate an SVG schematic. But it's a lot of work to define the same netlist twice with different types (`Part` for SKIDL, `V`/`R`/etc. for PySpice).

Is there any way to do both things from the same schematic?

**Comment by devbisme** [comment]
Here's one way to write some SKiDL code that generates an SVG schematic or does some SPICE stuff:

```py
from skidl.pyspice import *

draw_schematic = False  # Set True to draw schematics, False to do SPICE stuff.

if draw_schematic:
	# Use the KICAD Device library if you want to draw schematics.
	uselib = SchLib("Device", tool=KICAD)
	set_default_tool(KICAD)
else:
	# Otherwise, use the PySpice library.
	uselib = SchLib("pyspice", tool=SKIDL)
	set_default_tool(SPICE)

# Create your circuit using the selected library of parts.
r = Part(uselib, "R", value=1@u_kOhm)
c = Part(uselib, "C", value=1@u_uF)
gnd & r & c & gnd

if draw_schematic:
	# Draw the schematic.
	generate_svg()
else:
	# Otherwise, do some SPICE stuff in here...
	print(generate_netlist())

```
You'll probably want a custom library of SPICE symbols when generating a schematic since the `Device` library won't have things like voltage and current sources. So you could create that using the KICAD symbol editor (plus reusing some existing KICAD symbols for capacitors, resistors, etc.). Once you had that, you could use the SPICE part shorthand (`R`, `C`, etc.) for defining your circuit like so:

```py
from skidl.pyspice import *

draw_schematic = False

if draw_schematic:
	# Use the KICAD Device library if you want to draw schematics.
	uselib = SchLib("My_SPICE_Library", tool=KICAD)
	# Import the part names into the namespace so they can be instantiated easily.
	import sys
	_this_module = sys.modules[__name__]
	for p in uselib.get_parts():
		setattr(_this_module, p.name)
	set_default_tool(KICAD)
else:
	# Otherwise, use the PySpice library.
	uselib = SchLib("pyspice", tool=SKIDL)
	set_default_tool(SPICE)

# Create your circuit using the selected library of parts.
r = R(value=1@u_kOhm)
c = C(value=1@u_uF)
gnd & r & c & gnd

if draw_schematic:
	# Draw the schematic.
	generate_svg()
else:
	# Otherwise, do some SPICE stuff in here...
	print(generate_netlist())
```

**Reply by gsingh93** [reply]
Thanks! Since KiCad has some SPICE symbol libraries built-in (`pspice` and `Simulation_SPICE`), would it be possible to use those from SKIDL/PySpice? Based on `pyspice_sklib.py`, I see that to make a SKIDL part compatible with PySpice you have to pass the `Part` constructor a `pyspice` field that's pretty simple. When parsing the KiCad libraries, would it be possible to add their SPICE information to this field?

On a related note, I achieved what I wanted another way, by adding the `pyspice` field manually to components from the KiCad `Device` library and renaming some pins:
```python
from skidl.pyspice import *

draw_schematic = True

set_default_tool(KICAD)

vcc = Part("Device", "Battery", value = 5 @ u_V)
r1 = Part("Device", 'R', value=1@u_kOhm)
r2 = Part("Device", 'R', value=2@u_kOhm)

vcc.pyspice = V.pyspice
vcc.p1.name = 'p'
vcc.p2.name = 'n'

r1.pyspice = R.pyspice
r1.p1.name = 'p'
r1.p2.name = 'n'

r2.pyspice = R.pyspice
r2.p1.name = 'p'
r2.p2.name = 'n'

vin, vout, gnd = Net('Vin'), Net('Vout'), Net('GND')
vin.netio = 'i'
vout.netio = 'o'
gnd.netio = 'o'

gnd & vcc['n p'] & vin & r1 & vout & r2 & gnd

if draw_schematic:
    generate_svg()
else:
    set_default_tool(SPICE)
    circ = generate_netlist()
    print(circ)
    sim = circ.simulator()
    analysis = sim.dc(V1=slice(0, 5, 0.5))

    dc_vin = analysis.Vin
    dc_vout = analysis.Vout

    print('{:^7s}{:^7s}'.format('Vin (V)', ' Vout (V)'))
    print('='*15)
    for v, i in zip(dc_vin.as_ndarray(), dc_vout.as_ndarray()):
        print('{:6.2f} {:6.2f}'.format(v, i))
```

I'm wondering if there's any way to clean this approach up a bit and integrate it into SKIDL, but I'm not familiar enough with SKIDL or PySpice to know if this is a good idea. I was thinking the library could have some helper function that takes a non-PySpice part and makes it a PySpice part by doing what I did above:
```python
r = make_pyspice_resistor(Part("Device", "R", dest=TEMPLATE))
```

What do you think about this?

**Reply by gsingh93** [reply]
Any thoughts on this @devbisme?

**Reply by devbisme** [reply]
Apologies for my late reply. I hate SPICE. I hate SKiDL's SPICE interface. I hate looking at it. I hate thinking about it.

That said, we should try to support KiCad's SPICE elements. The `pspice.lib` file doesn't look like it contains any SPICE directives. I guess we would have to infer the appropriate elements from the F0 and F1 fields.

The parts in `Simulation_SPICE.lib` do have SPICE fields. We would need to code some translation from that information to create the `pyspice` dictionary that goes into a SKiDL part.

I did notice that both parts include an OPAMP. No model is included for it and `Simulation_SPICE` lists it as a SPICE primitive. NGspice doesn't have an OPAMP primitive, so I'm not sure how this part fits.

I like your technique for borrowing the `pyspice` field and adding it to non-SPICE parts. I would not have though to do that. I think we could write a simple conversion function to do it:

```py
r = conv_to_spice(R, Part("Device", "R", dest=TEMPLATE))
```
This function would copy the `pyspice` field from the SPICE component to the non-SPICE component. Then it would go through the pins and assign the SPICE component pin name to the pin in the non-SPICE component that had the same pin number. This would work as long as the pin numbers have the same function in both parts. That might not be true for things like transistors which assign E, B, and C to various combinations of pins 1, 2, and 3. We might have to compare pin names with their replacements to flag any possible incompatible assignments.

Of course, this doesn't work with XSPICE-type parts that include a MODEL. I'd have to think about that more to see if there's a simple way to do it.

**Reply by devbisme** [reply, marked as answer]
I added a simple method to `Part` objects to convert them into a form that's compatible with SPICE:

```py
def convert_for_spice(self, spice_part, pin_map):
    """Convert a Part object for use with SPICE.

    Args:
        spice_part (Part): The type of SPICE Part to be converted to.
        pin_map (dict): Dict with pin numbers/names of self as keys and num/names of spice_part pins as replacement values.
    """
```
Here's a simple test program I wrote that shows how it's used:
```py
from skidl.pyspice import *

def test_part_convert_for_spice():

    set_default_tool(KICAD)

    vcc = Part("Device", "Battery", value=5 @ u_V)
    r1 = Part("Device", "R", value=1 @ u_kOhm)
    r2 = Part("Device", "R", value=2 @ u_kOhm)

    vcc.convert_for_spice(V, {1: "p", 2: "n"})
    r1.convert_for_spice(R, {1: "p", 2: "n"})
    r2.convert_for_spice(R, {1: "p", 2: "n"})

    vin, vout, gnd = Net("Vin"), Net("Vout"), Net("GND")
    vin.netio = "i"
    vout.netio = "o"
    gnd.netio = "o"

    gnd & vcc["n p"] & vin & r1 & vout & r2 & gnd

    generate_svg()

    set_default_tool(SPICE)
    circ = generate_netlist()
    print(circ)
    sim = circ.simulator()
    analysis = sim.dc(V1=slice(0, 5, 0.5))

    dc_vin = analysis.Vin
    dc_vout = analysis.Vout

    print("{:^7s}{:^7s}".format("Vin (V)", " Vout (V)"))
    print("=" * 15)
    for v, i in zip(dc_vin.as_ndarray(), dc_vout.as_ndarray()):
        print("{:6.2f} {:6.2f}".format(v, i))
```
Right now, you'll have to install from the development branch to get this.

**Reply by Geocali** [reply]
Hello

I just started using Skidl, and in this example

```
from skidl.pyspice import *

draw_schematic = False  # Set True to draw schematics, False to do SPICE stuff.

if draw_schematic:
	# Use the KICAD Device library if you want to draw schematics.
	uselib = SchLib("Device", tool=KICAD)
	set_default_tool(KICAD)
else:
	# Otherwise, use the PySpice library.
	uselib = SchLib("pyspice", tool=SKIDL)
	set_default_tool(SPICE)

# Create your circuit using the selected library of parts.
r = Part(uselib, "R", value=1@u_kOhm)
c = Part(uselib, "C", value=1@u_uF)
gnd & r & c & gnd

if draw_schematic:
	# Draw the schematic.
	generate_svg()
else:
	# Otherwise, do some SPICE stuff in here...
	print(generate_netlist())
```

, I run into this error:

```
  File "/home/user/.local/lib/python3.8/site-packages/skidl/tools/kicad/kicad.py", line 1608, in gen_svg_comp
    for unit in range(1, num_units + 1):
UnboundLocalError: local variable 'num_units' referenced before assignment
```

I am using the libraries of Kicad 6 (/usr/share/kicad/symbols)

Could you tell me if I missed something ?

**Reply by devbisme** [reply]
I've made multiple attempts but failed to reproduce this error. I assume you're using the version of SKiDL installed from PyPi and running under linux.

Maybe we should try a simpler example and see if that works for you:
```python
from skidl.pyspice import *

r = Part("pyspice", "R", value=1@u_kOhm)
c = Part("pyspice", "C", value=1@u_uF)
gnd & r & c & gnd

print(generate_netlist())
```
